src/persist-client/src/operators/shard_source.rs

```diff
@@ -66,27 +67,28 @@ use crate::{PersistLocation, ShardId};
 /// using [`timely::dataflow::operators::generic::operator::empty`].
 ///
 /// [advanced by]: differential_dataflow::lattice::Lattice::advance_by
-pub fn shard_source<K, V, D, F, G>(
+pub fn shard_source<T, K, V, D, F, G>(
     scope: &mut G,
     name: &str,
     clients: Arc<PersistClientCache>,
     location: PersistLocation,
     shard_id: ShardId,
-    as_of: Option<Antichain<G::Timestamp>>,
-    until: Antichain<G::Timestamp>,
+    as_of: Option<Antichain<T>>,
+    until: Antichain<T>,
     flow_control: Option<FlowControl<G>>,
+    granular_control: Option<GranularFlowControl<G>>,
```

Comment:
Why do we have both granular and normal flow control?

Reply:
This is a good question. Compute wants a way to backpressure at a timestamp level only right? Its unclear how we can merge these...